File: crack.c
#define CHARNUM 52              // passwords use permutes of 52 chars in total
#define _XOPEN_SOURCE
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <unistd.h>

int main(int argc, string argv[])
{
    if (argc != 2)              // WRONG USAGE
    {
        printf("Usage: ./crack <hash>\n");
        return 1;
    }

    string hash = argv[1];      // RIGHT USAGE
    char salt[3];
    strncpy(salt, hash, 2);
    salt[2] = '\0';                     // hash & salt (first 2 chars of hash)

    char charArr[CHARNUM];
    for (int i = 0, j = 122; i < CHARNUM; i++, j--)
    {
        if (j == 96)
        {
            j = 90;
        }
        charArr[i] = (char) j;
    }                                   // initialize array of 52 chars

    for (int n = 1; n < 6; n++)         // brute force: try all one-, two-, ... char passwords
    {
        int count[n];                       // array of int to generate string
        char pw[n + 1];                     // one extra byte for null terminator
        for (int i = 0; i < n; i++)
        {
            count[i] = 0;                   // [ , ] -> [0, 0]
            pw[i] = charArr[0];             // [0, 0] -> "AA"
        }
        pw[n] = '\0';                       // "AA\0"
        int countSum = 0;                   // check if [51, 51] is reached

        while (strcmp(crypt(pw, salt), hash) != 0 && countSum < 51 * n)
        {
            count[0]++;                     // while haven't found pw
            countSum = 0;
            for (int i = 0; i < n; i++)         // check for overflowing like [52, 0]
            {
                if (count[i] > 51)
                {
                    if (i < n - 1)
                    {
                        count[i + 1]++;
                        count[i] = 0;
                    }
                }

                pw[i] = charArr[count[i]];      // generate new string
                countSum += count[i];
            }                                   // NOTES: does not check [51, 51]
        }

        if (strcmp(crypt(pw, salt), hash) == 0)
        {
            printf("Password is: %s\n", pw);
            return 0;
        }                                   // check [51, 51], or if pw has been found
    }
}

File: crack.py
import crypt
import sys

# wrong usage
if len(sys.argv) != 2:
    print("Usage: python crack.py [hash]")
    sys.exit(1)

# hash & salt
h = sys.argv[1]
s = h[:2]

# init char list
CHARNUM = 52
charArr = []
ch = 122
for i in range(CHARNUM):
    charArr.append(chr(ch))
    ch -= 1
    if ch == 96:
        ch = 90

# try every length from 1 to 5 chars
for length in range(1, 6):

    # count list and password
    count = []
    pw = ""

    # initial password
    for i in range(length):
        count.append(0)
        pw += charArr[count[i]]

    # while pw not found and last pw for this length not reached
    while (crypt.crypt(pw, s) != h and not all(x == 51 for x in count)):

        # try next char
        count[0] += 1

        # reset pw
        pw = ""

        for i in range(length):

            # roll-over
            if count[i] > 51:
                if i < length - 1:
                    count[i + 1] += 1
                    count[i] = 0

            # append new char to pw
            pw += charArr[count[i]]

    # pw is found
    if crypt.crypt(pw, s) == h:
        print("Password is: " + pw)
        sys.exit(0)